### ISSUE TITLE
qlora fsdp: Must flatten tensors with uniform dtype but got torch.float32 and torch.bloat16

### ISSUE THREAD

**Issue by Sehyo** [axolotl-ai-cloud/axolotl#1482]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Flattening should occur correctly.

### Current behaviour

When trying the example fsdp qlora config, with 2x4090s, the error "Must flatten tensors with uniform dtype but got torch.float32 and torch.bloat16" appears and training fails to start. I tried disabling b16 in the config, but similar error message occurs instead in that case.

### Steps to reproduce

System 2x4090, torch version 2.2.2, latest upstream axolotl commit at time of posting.
CUDA_VISIBLE_DEVICES=1,3 accelerate launch -m axolotl.cli.train ./qlora-fsdp.yml
-> error

### Config yaml

```yaml
base_model: NousResearch/Llama-2-7b-hf
model_type: LlamaForCausalLM
tokenizer_type: LlamaTokenizer

load_in_8bit: false
load_in_4bit: true
strict: false

datasets:
    - path: super_manager_data.jsonl
      type: sharegpt
      conversation: chatml
      field_system: system
      field_human: user
      field_model: assistant

dataset_prepared_path: last_run_prepared
val_set_size: 0.00
output_dir: ./qlora-out

adapter: qlora
lora_model_dir:

sequence_len: 4096
sample_packing: true
pad_to_sequence_len: true

lora_r: 64
lora_alpha: 32
lora_dropout: 0.05
lora_target_modules:
lora_target_linear: true
lora_fan_in_fan_out:

lora_modules_to_save:
    - lm_head
    - embed_tokens

wandb_project: SNUSNUSuperManager
wandb_entity: 
wandb_watch:
wandb_name: 34b-fsdp-qlora-super-manager-001
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 4
num_epochs: 2
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 0.00001

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 0
eval_table_size:
saves_per_epoch: 1
debug:
deepspeed:
weight_decay: 0.0
fsdp:
  - full_shard
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: true
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_state_dict_type: SHARDED_STATE_DICT
special_tokens:
    bos_token: "<|im_start|>"
    eos_token: "<|im_end|>"
    unk_token: "<unk>"
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10.12

### axolotl branch-commit

main/bf4cd672522729e5b0ad9a5523ad83a60de19243

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Duplicate of #1426
